pkg/certagent/agent.go

```diff
@@ -108,17 +111,31 @@ func GenerateCSRObject(config CSRConfig) (*capi.CertificateSigningRequest, error
 
 // RequestCertificate will create a certificate signing request for a node
 // with the config given and send it to a signer via a POST request.
-// If something goes wrong it returns an error.
+// If something goes wrong it returns an error unless CSRConfig.MaxRetry
+// is defined in which case we will retry the request.
 // NOTE: This method does not return the approved CSR from the signer.
 func (c *CertAgent) RequestCertificate() error {
 	csr, err := GenerateCSRObject(c.config)
 	if err != nil {
 		return fmt.Errorf("error generating CSR Object: %v", err)
 	}
 
-	// send CSR to the signer
-	if _, err := c.client.Create(csr); err != nil {
-		return fmt.Errorf("error sending CSR to signer: %v", err)
+	backoff := wait.Backoff{
+		Steps:    c.config.MaxRetry,
+		Duration: 10 * time.Second,
+		Factor:   0.0,
+	}
+	// implement a retry mechanism in the case request fails.
+	errc := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		// send CSR to the signer
+		_, err := c.client.Create(csr)
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if errc != nil {
+		return fmt.Errorf("error sending CSR to signer: %v", errc)
```

Comment:
errc is always going to be `wair.ErrTimeout`...
try something like this:

```go
var lastErr error
errc := wait.ExponentialBackoff(backoff, func() (bool, error) {
	// send CSR to the signer
	_, err := c.client.Create(csr)
	if err != nil {
                lastErr = err
		return false, nil
	}
	return true, nil
})
if errc != nil {
        if errc == wait.ErrTimeout && lastErr != nil {
            errc = lastErr
        }
	 return fmt.Errorf("error sending CSR to signer: %v", errc)
}
```

Reply:
Ok yeah, that makes sense `wait.ErrTimeout` is nice!